### PR TITLE
Turn non-200 HTTP responses into an error instead of trying to parse the body as a JSON-RPC response

### PR DESCRIPTION
**Pull request by romac**
Closes: #1359

This PR checks for the status code to be 200 OK, if not it returns an error with the actual status code and the body of the response, but does not attempt to parse the response body as a JSON-RPC payload.

CometBFT will [always set the status code](https://github.com/cometbft/cometbft/blob/86c07f20f163c17a0f5b7149cbd441a242b88557/rpc/jsonrpc/server/http_server.go#L151) to 200 OK and the content type to JSON for response which were successfully handled (which does mean that the JSON-RPC response itself is successful).

* [x] Referenced an issue explaining the need for the change
* [x] Updated all relevant documentation in docs
* [x] Updated all code comments where relevant
* [ ] Wrote tests
* [x] Added entry in `.changelog/`

### CODE REVIEW

File: rpc/src/client/transport/http.rs
Comment (mzabaluev) [inline rpc/src/client/transport/http.rs:375]: I'd rather put out a single tracing event, using keywords:

```rust
    tracing::debug!(status = %response_status, body = %response_body, "incoming response");
```

On the other hand, is it too much spam for debug level? Maybe the body dump should be on `trace` level.

File: rpc/src/client/transport/http.rs
Comment (mzabaluev) [inline rpc/src/client/transport/http.rs:381]: There is a method `StatusCode::is_success`, but maybe success statuses that are not exactly 200 should also be seen as non-conformant.

File: rpc/src/error.rs
Comment (mzabaluev) [inline rpc/src/error.rs:98]: Is there ever a need to dump the whole response body (which could be arbitrarily large) in error handling? Please leave it to the logs.

File: rpc/src/client/transport/http.rs
Comment (romac) [reply rpc/src/client/transport/http.rs]: Yes it's my understanding from reading both the Comet source and [this spec](https://www.simple-is-better.org/json-rpc/transport_http.html), that only 200 OK responses can be interpreted as a valid JSON-RPC payload. That said, I don't know how official that spec is.

Also it does not deal with redirects, etc. but that was already the case before. Do you know if hyper takes care of those for us?

File: rpc/src/error.rs
Comment (romac) [reply rpc/src/error.rs:98]: Good point!

Comment (romac) [reply rpc/src/error.rs]: [`2dc92cc`](https://github.com/informalsystems/tendermint-rs/pull/1361/commits/2dc92cc35a6beb8afca2c9dd6dd99f8eee178ed0)

File: rpc/src/client/transport/http.rs
Comment (romac) [reply rpc/src/client/transport/http.rs:376]: Yeah let's do that. I think it's fine at the debug level, and we're using it as via from Hermes `--debug=rpc` already, so let's leave it like that.

Comment (romac) [reply rpc/src/client/transport/http.rs]: e64df0b3

**Review by mzabaluev — APPROVED** [review]

**Comment by mzabaluev** [comment]
Conditional compilation needs to be fixed for wasm, otherwise looks fine.

**Comment by romac** [comment]
I will do this after we merge #1362

**Comment by romac** [comment]
> Conditional compilation needs to be fixed for wasm, otherwise looks fine.

https://github.com/informalsystems/tendermint-rs/pull/1361/commits/0b0f4c482b67b13357b1f0ba40acecb804829f33

**Comment by romac** [comment]
I've tested this from Hermes, and it works as expected.

**Comment by mzabaluev** [comment]
> Conditional compilation needs to be fixed for wasm, otherwise looks fine.

In fact, why do we build RPC stuff for wasm?

**Comment by romac** [comment]
> In fact, why do we build RPC stuff for wasm?

Because one can define a `Client` implementation which uses eg. the browser's `fetch` method and then compile all that to WASM.

**Pull request merged**
